### PR TITLE
Add localized LED flash effect on keypress to Massdrop ALT

### PR DESCRIPTION
**Pull request by reywood**
Hi there. This pull request evolved from trying to implement some custom LED effects on the Massdrop ALT. The code here causes a flash to ripple out from a key's LED when pressed. I had to modify the arm_atsam code to make this possible, though I tried to make as few changes to it as possible, only enough to make more diverse types of effects possible. As such, I'd be interested in hearing from @patrickmt on these changes. [Here's a video](https://imgur.com/Ea6av2j) of the effect.

### CODE REVIEW

**Comment by drashna** [comment]
Would it be possible to rename the functions to match `https://docs.qmk.fm/#/feature_rgb_matrix?id=custom-layer-effects` type naming.

This is mostly as a precursor to start migrating the code to the QMK RGB Matrix feature, so we don't have two matrix functions.

File: tmk_core/protocol/arm_atsam/led_matrix.h [hunk 1]
```diff
@@ -86,6 +86,7 @@ typedef struct led_disp_s {
 } led_disp_t;
 
 uint8_t led_matrix_init(void);
+void led_matix_init_user(void);
```
Comment (patrickmt) [inline tmk_core/protocol/arm_atsam/led_matrix.h]: Misspelled "matrix" in function name

File: keyboards/massdrop/alt/keymaps/reywood/led_matrix_user.c [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+static uint8_t calculate_new_color_component_value(uint8_t max, uint8_t min);
+static void calculate_new_led_boosts(uint8_t new_led_boosts[]);
+static uint8_t calculate_new_led_boost_at(int index);
+static uint8_t get_propagated_boost_from_neighbors(int led_position);
+static uint8_t get_led_boost_at_keypos(uint8_t row, uint8_t col);
+static void set_new_led_boosts(uint8_t* new_led_boosts);
+static uint8_t map_key_position_to_led_index(uint8_t col, uint8_t row);
+
+
+void led_matix_init_user(void) {
```
Comment (patrickmt) [inline keyboards/massdrop/alt/keymaps/reywood/led_matrix_user.c]: Misspelled "matrix" in function name

File: tmk_core/protocol/arm_atsam/led_matrix.c [hunk 3]
```diff
@@ -459,13 +461,19 @@ uint8_t led_matrix_init(void)
 
     //Run led matrix code once for initial LED coloring
     led_cur = 0;
-    led_matrix_run((led_setup_t*)led_setups[led_animation_id]);
+    led_matix_init_user();
```
Comment (patrickmt) [inline tmk_core/protocol/arm_atsam/led_matrix.c]: Misspelled "matrix" in function name

File: tmk_core/protocol/arm_atsam/led_matrix.c [hunk 4]
```diff
@@ -459,13 +461,19 @@ uint8_t led_matrix_init(void)
...
+    led_matrix_run();
 
     DBGC(DC_LED_MATRIX_INIT_COMPLETE);
 
     return 0;
 }
 
+__attribute__ ((weak))
+void led_matix_init_user(void) {
```
Comment (patrickmt) [inline tmk_core/protocol/arm_atsam/led_matrix.c]: Misspelled "matrix" in function name

**Comment by patrickmt** [comment]
Looks nice! I have your keymap running on an ALT here. I made a few comments on "matrix" being misspelled in user init function. Also the default keymap works as expected for ALT and CTRL keyboards. I checked the changes on Massdrop's fork for the LED code and looks like there would be no conflicts there either.

File: tmk_core/protocol/arm_atsam/led_matrix.c (hunk 3 above)
Comment (drashna) [reply tmk_core/protocol/arm_atsam/led_matrix.c]:
```suggestion
    led_matrix_init_user();
```

File: keyboards/massdrop/alt/keymaps/reywood/led_matrix_user.c (hunk 2 above)
Comment (drashna) [reply keyboards/massdrop/alt/keymaps/reywood/led_matrix_user.c]:
```suggestion
void led_matrix_init_user(void) {
```

File: tmk_core/protocol/arm_atsam/led_matrix.c (hunk 4 above)
Comment (drashna) [reply tmk_core/protocol/arm_atsam/led_matrix.c]:
```suggestion
void led_matrix_init_user(void) {
```

File: tmk_core/protocol/arm_atsam/led_matrix.h (hunk 1 above)
Comment (drashna) [reply tmk_core/protocol/arm_atsam/led_matrix.h]:
```suggestion
void led_matrix_init_user(void);
```

**Comment by reywood** [comment]
@patrickmt Good catch on that typo!

@drashna Considering your comment about renaming, should I rename `led_mat[r]ix_init_user` to something like `rgb_matrix_init_user`? Is there anything else in particular you want me to rename?

**Comment by drashna** [comment]
I would say that `rgb_matrix_init_user` would be the better name, I think.  Any input on that @patrickmt?

**Comment by patrickmt** [comment]
Getting it more aligned with QMK is good, so `rgb_matrix_init_user` is fine by me!

**Comment by reywood** [comment]
Alright, I updated the function name to `rgb_matrix_init_user` and updated my file names to match. I didn't update the other functions that were already present in the `arm_atsam` code. I could update those as well, but don't want to step on any work that @patrickmt may already have in progress.

**Comment by drashna** [comment]
@reywood would you mind making a PR with the same changes to the CTRL board too?

**Comment by reywood** [comment]
I'll take a look. I don't have a CTRL to test on though, so no guarantees that the code will work 😬

**Comment by patrickmt** [comment]
I can take care of the testing if you do take this on.

**Comment by valen214** [comment]
@reywood I have been working with similar effect recently. It is kind of funny because I am doing this with CTRL.
My implementation only requires to modify `[keymap.c](https://github.com/valen214/Massdrop-Ctrl-Keyboard-Ripple-LED/blob/master/keyboards/massdrop/ctrl/keymaps/custom/keymap.c)` and nothing else.

I think I might take some time to have a look on your code if I can get something new.
BTW, would you mind to help me testing out my code with ALT keyboard? (just copy the `keymap.c` file to `./keyboards/massdrop/alt/keymaps/<profile name>` and build is good enough, I hope)

**Pull request merged**